tensorflow/core/common_runtime/dml/dml_pooled_heap.cc

```diff
@@ -196,6 +206,12 @@ void DmlPooledHeap::ReclaimAllocations() {
   }
 }
 
+void DmlPooledHeap::OnDeviceRemoved() {
+  chunks_.clear();
+  total_capacity_ = 0;
```

Comment:
If this method gets called from a callback, what guarantees thread-safety here? Both the readback heap and upload heap are meant to be thread-safe.

Reply:
(Similar comment with the DeviceRemoved() method - there's nothing to guarantee that device_removed_ isn't being written to at the same time DeviceRemoved() is trying to read from it)

---
In reply to: [498592247](https://github.com/microsoft/tensorflow-directml/pull/64#discussion_r498592247) [](ancestors = 498592247)

Reply:
Well, it's called from the execution context's `Flush()`, which has a lock. Nothing stops external callers from calling `Trim()` from thread-unsafe code either, but the caller shouldn't do that.

`DeviceRemoved()` is a protected function that is only called under locks as well. There's no way to call it externally.

Reply:
The execution context has a lock, but the pooled heap doesn't execute underneath that lock (the upload heap and readback heap maintain their own independent locks). What stops one thread from calling into the execution context and triggering a device removal, simultaneous with a thread calling into the upload/readback heap and modifying this internal state?

---
In reply to: [498945010](https://github.com/microsoft/tensorflow-directml/pull/64#discussion_r498945010) [](ancestors = 498945010)

Reply:
Good point, I didn't think about this scenario. `OnDeviceRemoved()` should acquire the lock as well.